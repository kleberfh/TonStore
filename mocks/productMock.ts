import {IProductCart} from "../types";

export const ProductExample: IProductCart = {
  id: 1,
  count: 1,
  price: 1000,
  title: 'Product 1',
  features: ['feature1', 'feature2'],
  image: 'https://picsum.photos/200/300',
};