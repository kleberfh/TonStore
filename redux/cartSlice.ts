import { createSlice } from '@reduxjs/toolkit';
import { RootState, ICartState } from '../types';

const initialState: ICartState = {
  count: 0,
  products: [],
  value: 0,
};

export const cartSlice = createSlice({
  name: 'cart',
  initialState,
  reducers: {
    addProduct: (state, action) => {
      const alreadyInCart = state.products.find(
        product => product.id === action.payload.id,
      );
      if (alreadyInCart) {
        state.products.map(product => {
          if (product.id === action.payload.id) {
            product.count += 1;
          }
          return product;
        });
      } else {
        state.products.push({ ...action.payload, count: 1 });
      }
      state.count += 1;
      state.value += action.payload.price;
    },
    removeProduct: (state, action) => {
      if (action.payload.count > 1) {
        state.products = state.products.map(product => {
          if (product.id === action.payload.id) {
            product.count -= 1;
          }
          return product;
        });
      } else {
        state.products = state.products.filter(
          product => product.id !== action.payload.id,
        );
      }
      state.count -= 1;
      state.value -= action.payload.price;
    },
    clearCart: state => {
      state.count = 0;
      state.value = 0;
      state.products = [];
    },
  },
});

export const getCartValue = (state: RootState) => state.cart.value;
export const getCartCount = (state: RootState) => state.cart.count;
export const getCartProducts = (state: RootState) => state.cart.products;

export const { addProduct, removeProduct, clearCart } = cartSlice.actions;

export default cartSlice.reducer;
