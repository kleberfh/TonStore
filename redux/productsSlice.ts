import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getProducts } from '../services/products';
import { RootState, IProductState } from '../types';

const initialState: IProductState = {
  count: 0,
  products: [],
  status: 'idle',
};

export const fetchProducts = createAsyncThunk('posts/fetchPosts', async () => {
  try {
    return await getProducts();
  } catch (e) {
    return e.message;
  }
});

export const productsSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {},
  extraReducers(builder) {
    builder
      .addCase(fetchProducts.pending, state => {
        state.status = 'loading';
      })
      .addCase(fetchProducts.fulfilled, (state, action) => {
        state.status = 'success';
        state.products = action.payload.products;
        state.count += action.payload.count;
      })
      .addCase(fetchProducts.rejected, state => {
        state.status = 'error';
      });
  },
});

export const getAllProducts = (state: RootState) => state.products.products;
export const getProductsStatus = (state: RootState) => state.products.status;

export default productsSlice.reducer;
