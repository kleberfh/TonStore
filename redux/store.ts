import { combineReducers, configureStore } from '@reduxjs/toolkit';
import productsSlice from './productsSlice';
import cartSlice from './cartSlice';

const rootReducer = combineReducers({
  cart: cartSlice,
  products: productsSlice,
});

export default configureStore({
  reducer: rootReducer,
});