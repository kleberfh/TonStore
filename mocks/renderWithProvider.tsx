import React from "react";
import {Provider} from "react-redux";
import store from "../redux/store";

export default function RenderWithProvider({ children }: { children: React.ReactNode }) {
  return (
    <Provider store={store}>
      {children}
    </Provider>
  );
}