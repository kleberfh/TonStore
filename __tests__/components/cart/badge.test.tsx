import React from "react";
import renderer from 'react-test-renderer';
import RenderWithProvider from "../../../mocks/renderWithProvider";

import store from "../../../redux/store";
import {ProductExample} from "../../../mocks/productMock";
import Badge from "../../../components/cart/badge";
import {addProduct} from "../../../redux/cartSlice";

describe("Cart Badge", () => {
  it("should not render when empty", () => {
    const tree = renderer.create(
      <RenderWithProvider>
        <Badge />
      </RenderWithProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shound render when counter has value", () => {
    store.dispatch(addProduct(ProductExample));
    const tree = renderer.create(
      <RenderWithProvider>
        <Badge />
      </RenderWithProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

