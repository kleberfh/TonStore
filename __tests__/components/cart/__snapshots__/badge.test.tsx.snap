// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Badge should not render when empty 1`] = `null`;

exports[`Cart Badge shound render when counter has value 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ff2450",
      "borderRadius": 40,
      "height": 20,
      "justifyContent": "center",
      "position": "absolute",
      "right": -10,
      "top": -5,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
      "width": 20,
      "zIndex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
        "fontSize": 14,
        "fontWeight": "bold",
      }
    }
  >
    1
  </Text>
</View>
`;
